### PR TITLE
feat(spans): apply PII redaction in the ingest pipeline

### PR DESCRIPTION
**Pull request by csansoon**
Wires the PII redaction engine from #4245 into the ingest pipeline. That PR landed the detectors and the span walk as an inert module with no callers; this one makes them run, gated per project. Phase 2 of `specs/pii-redaction.md`.

Redaction is still off for every project until someone sets `settings.redaction`, and there is no UI yet — that's Phase 3. With redaction off, inserted rows are byte-identical to today's output and the pass costs 0 ms, asserted by test.

## how the policy reaches the worker

The policy is resolved at the **ingest boundary**, not in the worker, and stamped onto the queue job.

`ingestSpansUseCase` already loads every resolved `Project` with its settings to make the sampling decision, so the project half of the org → project cascade costs no extra query. Only the organization half needs a lookup, and `resolveOrganizationRedactionCached` keeps that off the hot path (60 s TTL, key `org:${organizationId}:settings:redaction`). Deciding here also pins the decision to the moment of ingest, so a toggle flipped between enqueue and processing can't apply retroactively to a batch already accepted.

Projects resolving to `off` are omitted from the map, and the `redaction` field is dropped entirely when the map is empty. **A missing field means "redact nothing"**, which is what makes this deployable in either order: an old worker ignores the field, and a new worker seeing no field is correct, because at deploy time no project has opted in. No legacy fallback needed — contrast the `projectId` workaround at `span-ingestion.ts:67`, which exists because that field change wasn't fail-safe.

The organization setting is read in the ingest route rather than the use case, since the cached resolver is a platform concern and the domain must not reach into the platform layer.

## where redaction runs, and why that's the only place it needs to run

Between `decodeAndTransform` and `repo.insert` in `processIngestedSpansUseCase`.

That single point covers every content sink: `traces` and `sessions` are materialized views on `spans`, every other derived table (`trace_search_documents`, `memory_blobs`, `dataset_rows`, `taxonomy_*`, `session_moment_labels`) re-reads `spans`, and `TracesIngested` hasn't fired yet, so no downstream consumer — including the PostHog destination export, which ships raw messages to a third party — ever sees unredacted content.

I verified that claim structurally rather than inferring it, since it's the thing most worth doubting:

- `grep -rn 'table: "spans"'` → one site, in `span-repository.ts`
- `grep -rn 'repo.insert'` → one caller, `process-ingested-spans.ts`
- `grep -rn '"span-ingestion"'` → one publisher, one consumer
- seed tooling POSTs to `/v1/traces`, so it traverses the same path

## failure behavior

Fails closed. A redaction failure propagates and fails the job; it is deliberately *not* caught alongside `SpanDecodingError`, which is swallowed as a warning. Because redaction is non-retroactive and there is no delete path, a fail-open write would be permanent, so losing spans for an opted-in project is the better trade.

Worth knowing precisely what "fails the job" costs, because I got this wrong in the commit messages before checking: **`span-ingestion` has no retries.** The publish passes no `attempts` and `new Queue` sets no `defaultJobOptions`, so BullMQ's default of one attempt applies and a failed batch is dropped rather than retried. Fail-closed is still right — nothing unredacted is written either way — but the cost is a dropped batch, not a retried one. Whether ingest should retry first is raised as an open question in the spec rather than changed here, since adding `attempts` alters failure handling for all ingest. Duplicate inserts on retry would be safe: `spans` is a `ReplacingMergeTree`.

**A present but malformed policy fails the job** rather than being skipped. Skipping looks like the safe default but is the *less* redacting choice: it resolves a corrupt policy on a project that opted in to a plaintext write. Absent and malformed are handled differently on purpose.

## buffered payload cleanup

The `tmp-ingest` object holding the raw OTLP payload is deleted once the spans are durable. The object-store lifecycle rule (1 day) is now a backstop rather than the only bound — self-hosted disk backends have no lifecycle rule at all. A failed delete doesn't fail an already-successful ingest.

This narrows but does not close the pre-worker exposure: payloads ≤50 KB still ride inline in the BullMQ job body in Redis, and BullMQ retains 1000 completed and 1000 failed jobs. Failed jobs don't churn. Documented in the spec as T-3 rather than papered over.

## two changes to the engine from #4245

**Removed the dry-run mode.** It was indistinguishable from `off` to a customer: it redacted nothing and surfaced nothing they could read, since its only output was worker span annotations landing in our own telemetry. It still cost a full scan and could still fail closed on a deadline overrun, so it could lose spans that `off` would have kept. Removing it collapsed real plumbing — `RedactionPolicy` lost `mode` entirely (a policy's presence is now the decision), the `mutate` flag threaded through the walk is gone, and `countRedactions` went with it.

That removed a safety net, so two things are now true and recorded in the spec: a false negative is invisible to everyone, which raises the stakes on detector precision; and tuning a detector default belongs in an offline script over real spans, not a product mode.

**Fixed a silent miss in the credit-card detector.** The candidate pattern allowed an optional space or dash between any two digits, so it bridged a card into the number following it. In `4111111111111111 123` the bridged run is 19 digits, a length Visa does issue, so it matched, failed Luhn, and consumed the real card with it. Fixed the way the IBAN detector already was: separate compact and grouped patterns, the grouped ones backreferencing their own separator.

Worth knowing how that was found, because it says something about the test suite: I hit it correcting a test expectation, not from a failing test. My first regression vectors *didn't reproduce it* — a neighbour long enough to overflow the length gate just backtracks to the card. Only a trailing number that pads the run to a valid card length reproduces it. Confirmed by mutation-testing the corrected vectors.

## observability

Annotations on `spans.processIngestedSpans`: `redaction.spans`, `redaction.matches` plus `redaction.matches.<entity>`, `redaction.leavesScanned`, `redaction.charsScanned`, `redaction.droppedAttributeKeys`, `redaction.oversizedFields`, `redaction.pseudonymizedIdentities`, `redaction.identityFallback`. Oversized fields warn; the identity fallback logs at error level, since it silently changes what's stored.

## performance

Measured over 200 coding-agent-shaped spans (prose, a diff, a tool call carrying that diff) through the real transform, warm-up discarded, 40 batches sampled:

| scanned content | per-span p50 | per-span p99 |
| --- | --- | --- |
| ~8 KB/span | 0.153 ms | 0.182 ms |
| ~29 KB/span | 0.685 ms | 1.223 ms |

Added CPU at the 29 KB point is 0.590 ms/span, about 1,700 spans/sec per core. The §4.7 budget is ≤5 ms/span at ~32 KB, so there's roughly 4× headroom at p99.

Two notes, because "concurrency 50" invites a wrong reading: the worker event loop is single threaded, so concurrency interleaves batches rather than parallelising this cost, and projects with redaction off measure 0.000 ms because the pass returns before touching a span. The benchmark isn't committed — the repo has no benchmark convention and it would add ~15 s to every `@domain/spans` CI run; the methodology is recorded in the spec instead.

## testing

`pnpm typecheck` clean across all 90 packages, knip and format clean, 1241 `@domain/spans` + 168 `@domain/shared` + 318 `@platform/db-postgres` tests passing.

Three behaviors were mutation-tested rather than assumed: skipping malformed policies breaks 2 tests, removing the buffered-payload delete breaks 1, and reintroducing the old credit-card pattern breaks 3.

Heads up on one flake unrelated to this change: `@platform/db-postgres` has pre-existing PGlite `beforeAll` contention flakiness. The same tree gave 47/47 passing and 5 timed-out `setupTestPostgres` hooks on consecutive full runs, and the affected files pass in isolation.

## not in this PR

No UI, no public API exposure, no docs — Phase 3. No ML tier for names and addresses — Phase 4. No deletion path for already-stored content — Phase 5, and it remains the honest answer to "we already leaked", which redaction can't give.

Refs #4245

## Summary by CodeRabbit

* **New Features**
  * Added per-project PII redaction during span ingestion, including stamped redaction payloads and optional identity pseudonymization (via a new `LAT_REDACTION_PSEUDONYM_SECRET` setting).
  * Added cached organization redaction policy resolution with cache invalidation and graceful cache failure.
* **Behavior Changes**
  * Redaction mode is now **Off** or **Enforce** only (dry-run removed), with redaction reporting reflecting actual redactions.
  * Batch processing now decodes and applies incoming redaction policies before persisting spans; buffered payloads are best-effort cleaned up after success.
* **Bug Fixes**
  * Improved credit-card detection accuracy and added stronger coverage for separator “bridging” edge cases.
* **Tests / Docs**
  * Updated redaction and ingestion test coverage and refreshed the PII redaction spec and environment documentation.

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -133,6 +133,13 @@ LAT_INGEST_PORT=3002
 # LAT_DEV_DESTINATIONS_SAFETY_LAG_MS=30000
 # LAT_WORKERS_HEALTH_PORT=9090
 
+# Secret keying the HMAC that pseudonymizes user ids and emails, for projects whose
+# PII redaction policy sets `identities: pseudonymize`. Generate a unique value for
+# production (openssl rand -hex 32). Rotating it changes every pseudonym, so spans
+# ingested before and after cannot be grouped by user. Leaving it unset degrades
+# those identity fields to full redaction rather than blocking ingestion.
+# LAT_REDACTION_PSEUDONYM_SECRET=
```
Comment (andresgutgon) [inline .env.example]: So this is mandatory to work in development if want to work on PII no?

Comment (csansoon) [reply .env.example]: No, not mandatory. Content redaction runs with nothing set. The secret only applies to projects that opt into `identities: pseudonymize`, and even then an unset secret fully redacts those two fields and logs an error rather than blocking ingest.

The question did catch a real problem though. Every other secret in this file ships an uncommented dev value (`LAT_MASTER_ENCRYPTION_KEY`, `LAT_BETTER_AUTH_SECRET`) and mine was commented out, which reads as "required but unconfigured". Fixed in e32af31b5: it has a dev value now, so pseudonymization also works locally without a setup step, and the comment starts with "Optional".

Leaving this thread open in case you want to push back on committing a dev value. The trade-off is that self-hosters who copy this file verbatim share an HMAC key, so pseudonyms become guessable across installs. That is also true of `LAT_MASTER_ENCRYPTION_KEY`, which guards more sensitive material, so I followed the existing convention rather than inventing a stricter rule for the weaker secret.

File: packages/domain/spans/src/redaction/detectors.ts [hunk 2]
```diff
@@ -39,7 +39,16 @@ const E164_PHONE_PATTERN = /(?<![\w+])\+[1-9]\d{7,14}(?!\d)/g
...
+/**
+ * Compact and grouped forms are separate patterns, and the grouped ones backreference
+ * their separator, for the same reason the IBAN detector does: a single pattern
+ * allowing an optional space between any two digits bridges two adjacent numbers,
+ * consumes both greedily, fails the checksum, and never reconsiders the real card
+ * inside. `+14155552671 4111111111111111` lost the card that way.
+ */
+const CREDIT_CARD_COMPACT_PATTERN = /(?<![\d.])\d{13,19}(?![\d.])/g
+const CREDIT_CARD_GROUPED_PATTERN = /(?<![\d.])\d{4}([ -])\d{4}\1\d{4}\1\d{1,7}(?![\d.])/g
+const CREDIT_CARD_AMEX_GROUPED_PATTERN = /(?<![\d.])\d{4}([ -])\d{6}\1\d{5}(?![\d.])/g
```
[the remark this answers is not in the document]
Comment (csansoon) [reply packages/domain/spans/src/redaction/detectors.ts]: Confirmed and fixed in 1d798657a. `3056 930902 5904` is Luhn-valid with a valid issuer prefix and was detected before the split, so a formatted Diners number was staying in plaintext under a default-enabled detector.

Added the 4-6-4 and 4-4-4-4-3 shapes. I kept them enumerated rather than generalising to "groups of four to six digits", because an open-ended repetition brings back the bridging this split was fixing: it swallows a trailing group, overruns 19 digits, and fails the length gate with the card inside the discarded match. Tests now cover every grouping a real issuer prints, in both separators, since the gap was a shape nobody had written down.

File: packages/domain/spans/src/use-cases/process-ingested-spans.ts [hunk 3]
```diff
@@ -146,6 +233,15 @@ export const processIngestedSpansUseCase =
       const repo = yield* SpanRepository
       yield* repo.insert(persistedSpans)
 
+      // The buffered payload holds unredacted content. It is dropped as soon as the
+      // spans are durable, so the object-store lifecycle rule is a backstop rather
+      // than the only bound — self-hosted disk backends have no lifecycle rule at
+      // all. A failed delete must not fail an already-successful ingest.
+      if (input.fileKey) {
+        const disk = yield* StorageDisk
+        yield* Effect.ignore(deleteFromDisk(disk, input.fileKey))
```
[the remark this answers is not in the document]
Comment (csansoon) [reply packages/domain/spans/src/use-cases/process-ingested-spans.ts]: Fixed in 7af76c8c7. The delete now lives in one helper called from both success exits, so a batch whose spans are all rejected no longer leaves the payload behind.

Kept it out of a finalizer deliberately: a failed job may be redelivered and redelivery needs the payload, so the failure paths have to leave it alone.

[the remark this answers is not in the document]
Comment (csansoon) [reply packages/domain/spans/src/use-cases/process-ingested-spans.ts]: Already fixed in d5ac70abf, which landed after the commit you reviewed (7cf409703). Same reasoning: deleting before the publishes left a stalled batch inserted with `TracesIngested` never fired, losing trace-end, billing, and search indexing.

There is a domain test for the invariant now, asserting that a failed publish leaves the spans inserted and the payload present.

File: packages/domain/spans/src/use-cases/ingest-spans.ts [hunk 4]
```diff
@@ -233,6 +272,7 @@ export const ingestSpansUseCase = (
       defaultProjectId,
       projectIdBySlug,
       isSandbox: input.isSandbox ?? false,
+      ...(redactedProjects > 0 ? { redaction } : {}),
```
[the remark this answers is not in the document]
Comment (csansoon) [reply packages/domain/spans/src/use-cases/ingest-spans.ts:275]: Right on the facts, and I reached the same conclusion independently while tracing this: no `attempts` on the publish and no `defaultJobOptions` on the queue, so a failed batch is dropped rather than retried. I had claimed otherwise in a few commit messages and corrected that in 91433bb46.

Not changing it in this PR though. Adding `attempts` alters failure handling for every ingest failure, not just redaction, and it deserves its own decision with the retry semantics thought through: duplicate inserts are safe because `spans` is a `ReplacingMergeTree`, but duplicate `TracesIngested` publishes feed billing, so the dedupe story needs checking first. Raised as open question 5 in `specs/pii-redaction.md`.

Worth noting the leak guarantee holds either way. Fail-closed means nothing unredacted is written; retries only change how much data a transient failure costs.

File: specs/pii-redaction.md [hunk 5]
```diff
@@ -366,8 +370,9 @@ Why not resolve in the worker: it would add an uncached `SettingsReader` Postgre
...
+- 60 s TTL, Zod-validated cached payload, `cache.hit` span annotation, and an `invalidateOrganizationRedactionCache(organizationId)` export mirroring `invalidateEffectivePlanCache` (`:95-101`).
+- **A cache failure degrades to a database read. A database failure propagates.** An earlier draft of this spec had the row read degrade to "no org policy" on the theory that the org layer can only raise strictness. That was wrong: degrading lets a `locked` org policy fall back to a weaker project policy and write plaintext, which is the exact failure the design invariant exists to prevent. It also buys no availability, because project resolution on this path already hard-depends on Postgres (`ingest-spans.ts:146-149`; `RepositoryError` is already in the use case's error union) and the request fails regardless.
+- Cache the *absence* of a policy explicitly rather than as a bare `null`. Almost every organization has no policy, and if a cached absence were indistinguishable from a miss the cache would never serve the common case.
```
[the remark this answers is not in the document]
Comment (csansoon) [reply specs/pii-redaction.md:375]: Both fixed in a034156d2.

The testing-plan line was the more important one: it still described the fallback §4.5 explicitly removed, so it would have reintroduced exactly the locked-organization-to-plaintext path. It now asserts cache failure falls back to a database read, a database failure propagates, and a cached absence is served rather than re-queried.

Also corrected the `invalidateRedactionPolicyCache` name to `invalidateOrganizationRedactionCache`.

File: specs/pii-redaction.md [hunk 6]
```diff
@@ -400,7 +404,7 @@ Justification, and this is a deliberate divergence from both competitors:
 2. For the in-process deterministic tier, "failure" means a code bug, so fail-open reduces to *silently writing plaintext PII for a customer who explicitly asked us not to*.
 3. There is no delete path ([§2.5](#25-what-does-not-exist)) and redaction is non-retroactive, so a fail-open write is permanent and unremediable.
 
-**The cost is explicit and accepted:** a persistent redaction bug loses spans for opted-in projects rather than leaking their PII. That is the correct trade for a compliance control, it is loud (failed jobs, error logs, retry exhaustion), and the customer has two self-service escapes: set `mode` back to `off`, or to `dryRun`.
+**The cost is explicit and accepted:** a persistent redaction bug loses spans for opted-in projects rather than leaking their PII. That is the correct trade for a compliance control, it is loud (failed jobs, error logs, retry exhaustion), and the customer has a self-service escape: set `mode` back to `off`.
```
[the remark this answers is not in the document]
Comment (csansoon) [reply specs/pii-redaction.md]: Fixed in a034156d2. The escape now says it applies only while the effective policy is project-controlled, and that under a locked organization policy recovery means changing the organization policy, which only an owner can do.

File: packages/domain/spans/src/redaction/detectors.test.ts [hunk 7]
```diff
@@ -104,10 +104,67 @@ describe("credit_card detector", () => {
...
+    ["4-4-4-4 Mastercard", "5500 0055 5555 5559"],
+    ["4-4-4-4 Discover", "6011 1111 1111 1117"],
+    ["4-4-4-1 Visa 13", "4222 2222 2222 2"],
+    ["4-6-5 Amex", "3782 822463 10005"],
+    ["4-6-4 Diners", "3056 930902 5904"],
+  ])("detects the %s grouping", (_shape, value) => {
+    expect(found(`card ${value} charged`, "credit_card")).toEqual([value])
+    const dashed = value.replaceAll(" ", "-")
+    expect(found(`card ${dashed} charged`, "credit_card")).toEqual([dashed])
+  })
```
[the remark this answers is not in the document]
Comment (csansoon) [reply packages/domain/spans/src/redaction/detectors.test.ts:129]: Good catch, and it was the same gap that let the Diners shape regress: I added the pattern without a test. Covered in 6e26d3f8d.

The overlap you predicted is real. A 19-digit card contains a Luhn-valid 16-digit Visa in its first four groups, so both patterns match at the same offset. Leftmost-longest already resolves it to the full number, verified before writing the test, so this was a coverage gap rather than a live bug. The test pins both the raw matches and the redacted output, since picking the shorter match would leave the final three digits sitting next to a placeholder.

It needs its own case rather than a row in the shape table, because that table asserts exactly one match per shape and this shape legitimately produces two.

File: packages/domain/spans/src/redaction/detectors.ts [hunk 8]
```diff
@@ -39,7 +39,25 @@ const E164_PHONE_PATTERN = /(?<![\w+])\+[1-9]\d{7,14}(?!\d)/g
 // Separated NANP forms only: a bare ten-digit run is indistinguishable from the numeric ids in tool output.
 const NANP_PHONE_PATTERN = /(?<![\w.-])(?:\(\d{3}\) ?|\d{3}[-. ])\d{3}[-. ]\d{4}(?![\d.-])/g
 
-const CREDIT_CARD_PATTERN = /(?<![\d.])\d(?:[ -]?\d){11,18}(?![\d.])/g
+/**
+ * Compact and grouped forms are separate patterns, and each grouped one backreferences
+ * its separator, for the same reason the IBAN detector does: a single pattern allowing
+ * an optional separator between any two digits bridges two adjacent numbers, consumes
+ * both greedily, fails the checksum, and never reconsiders the real card inside.
+ * `4111111111111111 123` lost the card that way.
```
[the remark this answers is not in the document]
Comment (csansoon) [reply packages/domain/spans/src/redaction/detectors.ts]: Half of this I have done, half I am declining.

Done: the card-shaped string in the detector comment was illustration rather than fixture, so it is gone in 6e26d3f8d, replaced by a prose description. That was the one place the value was not doing any work.

Declining the test fixtures. These are the card networks published test numbers, issued specifically to be hardcoded in tests, and they identify no cardholder. A detector for card numbers cannot be tested without card-shaped input, so removing them would mean deleting the coverage that caught the Diners regression in the first place. I added a comment recording their provenance so the next reader does not have to wonder.

On configuring OpenGrep suppression: the repository does not run OpenGrep. There is no `.semgrep`/`.opengrep` config, nothing in `.github/workflows`, and no `.coderabbit.yaml`, so this appears to be your scanner rather than our CI, and it is not gating this PR. I would rather not add configuration for a tool we do not run. If we adopt it later, a fixture allowlist for this file is the right call and I will happily add it then.
